stdlib/dynarray.ml

```diff
@@ -962,6 +962,42 @@ let for_all p a =
   check_same_length "for_all" a ~length;
   res
 
+let exists2 p a1 a2 =
+  let Pack {arr = arr1; length = length1; dummy = dummy1} = a1 in
+  let Pack {arr = arr2; length = length2; dummy = dummy2} = a2 in
+  check_valid_length length1 arr1;
+  check_valid_length length2 arr2;
+  if length1 <> length2 then invalid_arg "Dynarray.exists2";
+  let rec loop p arr1 dummy1 arr2 dummy2 i length =
+    if i = length then false
+    else
+      p (unsafe_get arr1 ~dummy:dummy1 ~i ~length)
+        (unsafe_get arr2 ~dummy:dummy2 ~i ~length)
+      || loop p arr1 dummy1 arr2 dummy2 (i + 1) length
+  in
+  let res = loop p arr1 dummy1 arr2 dummy2 0 length1 in
+  check_same_length "exists2" a1 ~length:length1;
+  check_same_length "exists2" a2 ~length:length2;
+  res
+
+let for_all2 p a1 a2 =
+  let Pack {arr = arr1; length = length1; dummy = dummy1} = a1 in
+  let Pack {arr = arr2; length = length2; dummy = dummy2} = a2 in
+  check_valid_length length1 arr1;
+  check_valid_length length2 arr2;
+  if length1 <> length2 then invalid_arg "Dynarray.for_all2";
+  let rec loop p arr1 dummy1 arr2 dummy2 i length =
+    if i = length then true
+    else
+      p (unsafe_get arr1 ~dummy:dummy1 ~i ~length)
+        (unsafe_get arr2 ~dummy:dummy2 ~i ~length)
+      && loop p arr1 dummy1 arr2 dummy2 (i + 1) length
+  in
```

Comment:
I think this should either be turned into a tail-call with a boolean accumulator or be rewritten as an imperative loop.

Reply:
I don't understand your remark, I believe that the right-hand-side of shortcut boolean operators is a tail-position, so this should already be a tail call.

Reply:
> I believe that the right-hand-side of shortcut boolean operators is a tail-position,

If you don't believe but are sure then I'm happy to have learned a new thing today.

Reply:
I am never sure when an OCaml expert suggests something to the contrary, but yes in my mind `&&` and `||` have the same computational behavior as their expansion in terms of if-then-else, and the right-hand-side is in tail position. I did just check by adding an explicit `[@tailcall]` attribute in a simple example, and it is accepted.